File: backend/src/models/quote.ts
export enum MarketStatus {
    PRE_MARKET = 'PRE_MARKET',
    OPEN = 'OPEN',
    CLOSED = 'CLOSED',
    POST_MARKET = 'POST_MARKET',
    UNKNOWN = 'UNKNOWN'
}

export interface NormalizedQuote {
    instrumentId: string;
    lastPricePaise: number;
    previousClosePaise: number;
    openPaise: number;
    highPaise: number;
    lowPaise: number;
    volume: number;
    dataTimestamp: Date;
    retrievedAt: Date;
    marketStatus: MarketStatus;
    source: string;
    /** Nominal delay of the feed itself (e.g. 900 for Yahoo's 15-min NSE
     * delay). Freshness bands are judged net of this; absent means real-time. */
    feedDelaySec?: number;
}

export interface Candle {
    tradeDate: Date;
    openPaise: number;
    highPaise: number;
    lowPaise: number;
    closePaise: number;
    volume: number;
}

export interface InstrumentRef {
    instrumentId: string;
    symbol: string;
    exchange: string;
    name: string;
}

File: internal/model/quote.go
package model

import "time"

type MarketStatus string

const (
    StatusPreMarket  MarketStatus = "PRE_MARKET"
    StatusOpen       MarketStatus = "OPEN"
    StatusClosed     MarketStatus = "CLOSED"
    StatusPostMarket MarketStatus = "POST_MARKET"
    StatusUnknown    MarketStatus = "UNKNOWN"
)

type NormalizedQuote struct {
    InstrumentID       string       `json:"instrumentId"`
    LastPricePaise     int64        `json:"lastPricePaise"`
    PreviousClosePaise int64        `json:"previousClosePaise"`
    OpenPaise          int64        `json:"openPaise"`
    HighPaise          int64        `json:"highPaise"`
    LowPaise           int64        `json:"lowPaise"`
    Volume             int64        `json:"volume"`
    DataTimestamp      time.Time    `json:"dataTimestamp"`
    RetrievedAt        time.Time    `json:"retrievedAt"`
    MarketStatus       MarketStatus `json:"marketStatus"`
    Source             string       `json:"source"`
}

type Candle struct {
    TradeDate             time.Time
    OpenPaise, HighPaise, LowPaise, ClosePaise int64
    Volume                int64
}

type InstrumentRef struct {
    InstrumentID string
    Symbol       string
    Exchange     string
    Name         string
}
